File: LeetCode/Top_Amazon_Questions/Array/01.Two_Sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<pair<int,int> > p;
        p.reserve(nums.size());
        for(int i=0;i<nums.size();++i){
            p.push_back(make_pair(nums[i],i));
        }
        sort(p.begin(),p.end());
        int l = 0,r = p.size()-1;
        while(l<r){
            if(p[l].first+p[r].first==target){
                return {p[l].second,p[r].second};
            }else if(p[l].first+p[r].first>target){
                r--;
            }else{
                ++l;
            }
        }
        return {-1,-1};
    }
};

File: LeetCode/Problems_JAVA/01.TwoSum.java
class pair{
        int index;
        int value;
        public pair(int index,int value){
            this.index = index;
            this.value = value;
        }
    }
    class Compare{
        static void compare(pair arr[]){
            Arrays.sort(arr,new Comparator<pair>(){
                public int compare(pair p1,pair p2){
                    return  p1.value-p2.value;
                }
            });
        }
    }
class Solution {

    public int[] twoSum(int[] nums, int target) {
        pair arr[] = new pair[nums.length];
        for(int i=0;i<nums.length;++i){
            arr[i] = new pair(i,nums[i]);
        }
        Compare obj = new Compare();
        obj.compare(arr);
        for(pair p:arr){
            System.out.println(p.index+" "+p.value);
        }
        int l = 0;
        int r = nums.length-1;
        int ans[] = new int[2];
        while(l<r){
            if(arr[l].value+arr[r].value==target){
                ans[0] = arr[l].index;
                ans[1] = arr[r].index;
                break;
            }else if(arr[l].value+arr[r].value<target){
                l = l+1;
            }else{
                r = r-1;
            }
        }
        return ans;
    }
}
